actors/migration/top.go

```diff
@@ -17,65 +23,159 @@ type StateMigration interface {
 }
 
 type ActorMigration struct {
-	InCodeCID      cid.Cid
 	OutCodeCID     cid.Cid
 	StateMigration StateMigration
 }
 
-var migrations = []ActorMigration{ // nolint:varcheck,deadcode,unused
-	{
-		InCodeCID:      builtin0.AccountActorCodeID,
+var migrations = map[cid.Cid]ActorMigration{ // nolint:varcheck,deadcode,unused
+	builtin0.AccountActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.AccountActorCodeID,
 		StateMigration: &accountMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.CronActorCodeID,
+	builtin0.CronActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.CronActorCodeID,
 		StateMigration: &cronMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.InitActorCodeID,
+	builtin0.InitActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.InitActorCodeID,
 		StateMigration: &initMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.StorageMarketActorCodeID,
+	builtin0.StorageMarketActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.StorageMarketActorCodeID,
 		StateMigration: &marketMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.StorageMinerActorCodeID,
+	builtin0.StorageMinerActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.StorageMinerActorCodeID,
 		StateMigration: &minerMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.MultisigActorCodeID,
+	builtin0.MultisigActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.MultisigActorCodeID,
 		StateMigration: &multisigMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.PaymentChannelActorCodeID,
+	builtin0.PaymentChannelActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.PaymentChannelActorCodeID,
 		StateMigration: &paychMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.StoragePowerActorCodeID,
+	builtin0.StoragePowerActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.StoragePowerActorCodeID,
 		StateMigration: &powerMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.RewardActorCodeID,
+	builtin0.RewardActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.RewardActorCodeID,
 		StateMigration: &rewardMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.SystemActorCodeID,
+	builtin0.SystemActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.SystemActorCodeID,
 		StateMigration: &systemMigrator{},
 	},
-	{
-		InCodeCID:      builtin0.VerifiedRegistryActorCodeID,
+	builtin0.VerifiedRegistryActorCodeID: ActorMigration{
 		OutCodeCID:     builtin.VerifiedRegistryActorCodeID,
 		StateMigration: &verifregMigrator{},
 	},
 }
+
+// A phoenix tracks the unburning of funds
+type phoenix struct {
+	burntBalance abi.TokenAmount
+}
+
+func (p phoenix) load(ctx context.Context, actorsIn *states.TreeTop) error {
+	burntFundsActor, err := actorsIn.GetActor(ctx, builtin0.BurntFundsActorAddr)
```

Comment:
I don't know that this is where the funds will come from. All funds were unburnt in a previous migration, back to mining rewards wallets or something. This needs to be easily configurable.

Reply:
Glad you pointed this out.  I will investigate.

Reply:
From @whyrusleeping its ok as policy to do this from burnt funds.  The concern is that we really want to make sure burnt funds has enough to do all of these so if it is a possibility that burnt funds won't cover ip debt we should take from the miner reserve.